controllers/apply-next/form-router.js

```diff
@@ -33,92 +37,70 @@ function initFormRouter(formModel) {
 
     /**
      * Route: Start page
      */
+    async function renderStartPage(req, res, errors = []) {
+        const { startPage } = res.locals.form;
+        const applications = await applicationsService.getApplicationsForUser(req.user.userData.id, sessionKeys.form);
+
+        res.render(startPage.template, {
+            title: res.locals.formTitle,
+            applications: applications,
+            csrfToken: req.csrfToken(),
+            errors: errors
+        });
+    }
+
     router
         .route('/')
-        .all((req, res, next) => {
-            const { startPage } = res.locals.form;
-            if (!startPage) {
-                throw new Error('No startpage found');
-            }
-            if (!startPage.urlPath && !startPage.template) {
-                throw new Error('No valid startpage types found');
-            }
-            next();
-        })
-        .get(cached.noCache, async function(req, res) {
-            const { startPage } = res.locals.form;
-            const applications = await ApplicationService.getApplicationsForUser(
-                req.user.userData.id,
-                sessionKeys.form
-            );
-            if (startPage.template) {
-                res.render(startPage.template, {
-                    title: res.locals.form.title,
-                    applications: applications,
-                    csrfToken: req.csrfToken()
-                });
-            } else if (startPage.urlPath) {
-                res.redirect(localify(req.i18n.getLocale())(startPage.urlPath));
-            }
+        .get(async function(req, res) {
+            renderStartPage(req, res);
         })
-        .post(formModel.titleField.validator(formModel.titleField), (req, res) => {
-            const errors = validationResult(req);
-            if (errors.isEmpty()) {
-                // create app and proceed!
-                ApplicationService.createApplication({
+        .post(async (req, res) => {
+            try {
+                const application = applicationsService.createApplication({
                     userId: req.user.userData.id,
                     formId: sessionKeys.form,
+                    // @TODO: Validate the title field
                     title: req.body['application-title']
-                })
-                    .then(application => {
-                        res.redirect(`${req.baseUrl}/edit/${application.id}`);
-                    })
-                    .catch(error => {
-                        Raven.captureException(error);
-                        return renderError(error, req, res);
-                    });
-            } else {
-                const { startPage } = res.locals.form;
-                res.render(startPage.template, {
-                    title: res.locals.form.title,
-                    csrfToken: req.csrfToken(),
-                    errors: errors.array()
                 });
+                res.redirect(`${req.baseUrl}/edit/${application.id}`);
```

Comment:
Does this need to be in `then()` callback (or should it be `await applicationsService`...)? Otherwise isn't there a race condition where it might redirect before the application has been saved?

Reply:
Forgot the `await` keyword, thanks!